### PR TITLE
[feat]: [CCM-11754]: Perspective preferences

### PR DESCRIPTION
**Pull request by sdsingh07**
## Describe your changes
At present, we don’t have a functionality in the perspectives that allows users to filter taxes, discounts etc. which are provided by the CSP’s dashboards. This often results in cost discrepancies.

As part of our commitment to perspective preferences, we aim to bridge this gap by creating CCM perspectives that are functionally similar to those of the CSP's dashboards. Our goal is to ensure accurate cost analysis and provide additional features for a seamless user experience.

Changes:
- Settings CRUD listener to read redis event on setting's update.
- Add `viewPreferences` filed in Perspective CRUD and GraphQL APIs
- Aggregation changes in perspective BigQueries
- GCP Invoice Month Filter Support
- Migration to add modified `viewPreferences` field across all the perspectives.

Design doc: [Perspective Preferences - Design](https://harness.atlassian.net/wiki/spaces/CE/pages/21381677074/Perspective+Preferences+-+Design)

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50585)

### CODE REVIEW

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

File: 450-ce-views/src/main/java/io/harness/ccm/views/utils/DateUtils.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.time.LocalDateTime;
+import java.time.Month;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+@OwnedBy(CE)
+public class DateUtils {
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/utils/DateUtils.java:24]: Move this class to `490-ce-commons`

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/costdetails/RESTWrapperCostDetails.java [hunk 2]
```diff
@@ -151,9 +151,10 @@ public class RESTWrapperCostDetails {
 
     log.info("Aggregations: {}", aggregationList);
 
+    ViewPreferences viewPreferences = ViewPreferences.builder().build();
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/costdetails/RESTWrapperCostDetails.java:154]: Pass viewPreferences as null

File: 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewPreferenceServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+  private static final String AWS_LINE_ITEM_TYPE_FIELD_NAME = "AWS Line Item Type";
+  private static final String GCP_COST_TYPE_FIELD_NAME = "GCP Cost Type";
+
+  @Inject @Named("PRIVILEGED") private NGSettingsClient settingsClient;
+  @Inject private ViewParametersHelper viewParametersHelper;
+
+  private final LoadingCache<String, List<SettingResponseDTO>> settingsResponseCache =
+      Caffeine.newBuilder()
+          .maximumSize(5)
+          .expireAfterWrite(30, TimeUnit.SECONDS)
```
Comment (sdsingh07) [inline 450-ce-views/src/main/java/io/harness/ccm/views/service/impl/CEViewPreferenceServiceImpl.java:93]: Change it to 10 seconds.

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger uts

**Comment by sdsingh07** [comment]
trigger ut6, ut9

**Comment by sdsingh07** [comment]
trigger uts

**Review by svmav — APPROVED** [review]
> Reviewed 24 of 39 files at r1, 3 of 5 files at r3, 3 of 4 files at r4, 6 of 11 files at r6, 1 of 2 files at r7, 6 of 6 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50585)* status: :shipit: complete! ✓

**Comment by svmav** [comment]
trigger ut7

**Comment by svmav** [comment]
trigger ut2

**Comment by sdsingh07** [comment]
trigger uts

**Review by svmav — APPROVED** [review]
> Reviewed 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50585)* status: :shipit: complete! ✓

**Comment by sdsingh07** [comment]
trigger ut0

**Comment by sdsingh07** [comment]
trigger ut0

**Pull request merged**